### PR TITLE
Add additional reporting output file options

### PR DESCRIPTION
**Pull request by MauriceReeves-usds**
This PR allows us to create two other types of output files for our schemas, specifically in a data dictionary format so we can create a CSV or Excel XLSX file. The data dictionary file format is based on the format we are using on the monkeypox effort to communicate our schema structure to other teams in the CDC.

This also includes some changes and cleanup of the documentation code to collect common functionality into base classes.

Test Steps:
1. Build the application
2. Run `./prime generate-docs --output-dir build/xl --generate-xl --input-schema covid-19 --no-generate-markup` to generate an Excel document
3. Look in the `build/xl/` folder to locate the XSLX file and verify it's correct
4. Run `./prime generate-docs --output-dir build/csv --generate-csv --input-schema covid-19 --no-generate-markup` to generate a CSV document
5. Look in the `build/csv/` folder to location the CSV file and verify it's correct
6. Run unit tests
7. Run integration tests
8. Run `./prime test` 

## Changes
- Unifies logic into base classes so things are more clean in the documentation code
- Adds a CSV documentation factory generator
- Adds an Excel documentation factory generator
- Adds additional tests for some of the documentation logic
- Adds command line logic to include writing out CSV and Excel files

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [x] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

### CODE REVIEW

File: prime-router/src/main/kotlin/docgenerators/DocumentationFactory.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import java.nio.file.Files
+import java.nio.file.Paths
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+
+/** The base documentation factory with core logic */
+abstract class DocumentationFactory {
+    abstract val fileExtension: String
+
+    abstract fun getSchemaDocumentation(schema: Schema): Sequence<*>
```
Comment (MauriceReeves-usds) [inline prime-router/src/main/kotlin/docgenerators/DocumentationFactory.kt:13]: Using the `*` means that you can pass a sequence of unknown type in subclasses. This is different than using `Any` which is like using `object`

File: prime-router/src/main/kotlin/docgenerators/DocumentationFactory.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * A string based documentation factory that returns a sequence of strings,
+ * as opposed to the [TableBasedDocumentationFactory] that returns a sequence of a list of strings
+ * for generating things like CSV and Excel. This is used for Markdown and HTML which have a non-table
+ * layout
+ **/
+abstract class StringBasedDocumentationFactory : DocumentationFactory() {
+    abstract override fun getSchemaDocumentation(schema: Schema): Sequence<String>
+}
```
Comment (MauriceReeves-usds) [inline prime-router/src/main/kotlin/docgenerators/DocumentationFactory.kt:69]: Make the sequence for getting the schema documentation more specific.

File: prime-router/src/main/kotlin/docgenerators/DocumentationFactory.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+/** A documentation factory that is based on a table of data */
+abstract class TableBasedDocumentationFactory : DocumentationFactory() {
+    open val delimiter = "|"
+
+    /** Gets the headers for the documentation factory */
+    abstract fun getHeaders(): Array<String>
+
+    /** Override for the  */
+    abstract override fun getSchemaDocumentation(schema: Schema): Sequence<List<String>>
+}
```
Comment (MauriceReeves-usds) [inline prime-router/src/main/kotlin/docgenerators/DocumentationFactory.kt:80]: Here as well with specificity for the sequence.

**Comment by brick-green** [comment]
There are some fields listed in the schema that are not in the output file:

- natFlatFileField
- hhsGuidanceField
- reference
- hl7AOEQuestion

Why were those left out?

**Comment by MauriceReeves-usds** [comment]
@brick-green I was following the format that DCIPHER and the CDC was using to communicate about different schemas, and those fields were not part of those spreadsheets.

**Review by brick-green — APPROVED** [review]
lgtm

**Pull request merged**